subsys/bluetooth/controller/ll_sw/nordic/lll/lll_clock.c

```diff
@@ -19,61 +19,75 @@
 #define LFCLOCK_TIMEOUT_MS 500
 #define HFCLOCK_TIMEOUT_MS 2
 
-static void clock_ready(struct device *dev, clock_control_subsys_t subsys,
-			void *user_data);
+struct lll_clock_state {
+	struct onoff_client cli;
+	struct k_sem sem;
+};
 
-static struct device *dev;
+static struct lll_clock_state lf_state;
+static struct lll_clock_state hf_state;
+static atomic_val_t hf_refcnt;
 
-int lll_clock_init(void)
+static void clock_ready(struct onoff_manager *mgr, struct onoff_client *cli,
+			uint32_t state, int res)
+{
+	struct lll_clock_state *clk_state =
+		CONTAINER_OF(cli, struct lll_clock_state, cli);
+
+	k_sem_give(&clk_state->sem);
+}
+
+int clk_request(struct onoff_manager *mgr, struct lll_clock_state *state)
 {
 	int err;
 
-	dev = device_get_binding(DT_LABEL(DT_INST(0, nordic_nrf_clock)));
-	if (!dev) {
-		return -ENODEV;
+	k_sem_init(&state->sem, 0, 1);
+	sys_notify_init_callback(&state->cli.notify, clock_ready);
+	err = onoff_request(mgr, &state->cli);
+	if (err < 0) {
+		return err;
 	}
 
-	err = clock_control_on(dev, CLOCK_CONTROL_NRF_SUBSYS_LF);
-
-	return err;
+	return 0;
 }
 
-int lll_clock_wait(void)
+static int on_wait(struct lll_clock_state *state, uint32_t timeout)
 {
-	static bool done;
+	int err = k_sem_take(&state->sem, K_MSEC(timeout));
 
-	if (done) {
-		return 0;
+	if (err == 0) {
+		k_sem_give(&state->sem);
```

Comment:
Why give after a take?

Reply:
rewrote it, initially, i didn't want to add another request to lf clock but since it is permanently requested it is much easier to add new blocking request.